### ISSUE TITLE
Metadata about a slot's value - Not Applicable; Missing; etc.

### ISSUE THREAD

**Issue by ddooley** [linkml/linkml#178]
NCBI Biosample and other related specifications have a set of "metadata" states a field can be in which can accompany a slot's instance value (usually a null value), e.g. "Not Applicable; Missing; Not Collected; Not Provided; Restricted Access". Does LinkML have a recommended way (or slot) to handle this?  The linkml:status field seems to be about the slot's curation state (finalization) rather than metadata about field values?

**Comment by cmungall** [comment]
You mean like this:
https://ena-docs.readthedocs.io/en/latest/submit/samples/missing-values.html

An enum can be used to encode a missing value value set..

but I think the challenge here to be able to encode a type compositionally; e.g. a lat_long slot that takes as value a geolocation object OR a missing value enum?

Note I'm dodging this in the mixs linkml
https://github.com/cmungall/mixs-source

I think mixs is a bit underspecified here. If a field is marked as required (globally, or per package or checklist) is it a violation of this constraint if a missing value field is specified?

**Comment by ddooley** [comment]
Yes, the INSDC missing values are what we have in mind, but I get that there are different ways of handling this.  I'd think any field/slot could be augmented with its own enumeration of missing values; or a named list of them. They are often multiplexed into fields that can handle strings too, but that probably isn't something LinkML needs to be concerned about.  But if a slot is "required" AND has a missing values list, then any of those missing values would be acceptable.  That forces the metadata state to be answered if a field doesn't have content. No missing values = traditional required functionality on the field.

**Comment by cmungall** [comment]
In the specific case of mixs/insdc it seems if a field is marked mandatory
then it is mandatory to have a non missing value. But we should confirm.

In the general case,  this could be handled by the schema designer making a
dedicated composite class or union.. but maybe there is a more elegant way

**Comment by ddooley** [comment]
One oddity/design hassle is that these data-exchange specifications can allow a missing values (or as they are referenced in DataHarmonizer templates, "data status" null values) list in the shared data, but they have to be stripped out (and maybe put into a separate column) before they go into SQL databases for date and numeric and coded categorical fields.

**Comment by cmungall** [comment]
We could represent it like this:

```
  age:
      exactly_one_of:
        - range: decimal
        - range: MissingValueEnum
```

We could also have a generic transform of boolean range conditions to the relational model, there would be a few options here

 1. as you say, two separate columns in the sample table/class
 2. an age table/class, with a backreference to the sample table/class, and two slots/columns

**Comment by ddooley** [comment]
Allowing unions for ranges sounds great.  I guess the "generic transform" is set up to enable tabular datastructures to encode/decode what is in a particular LinkML slot definition?  So a generic transform is invoked for data import or export?

I'm guessing one slight issue that might arise in future is multi-select - ability to supply several categorical values, but only one missingValueEnum if no categoricals are selected.

**Comment by nlharris** [comment]
I wonder whether some of these requests have now been implemented?

exactly_one_of is also mentioned in "exactly_one_of with range: does not work as expected when using linkml:types #2283"
